### ISSUE TITLE
unixfd handling should be improved

### ISSUE THREAD

**Issue by KillingSpark** [KillingSpark/rustbus#36]
The marshal/unmarshal API currently expects unixfds to be indexes into some array of actual unixfds. This is very unergonomic for users that want to transfer unixfds in deep structures (or even arrays/maps of types containing them). A automatic way of converting fds to indexes and storing the fds in the array would be preferrable. libdbus does this, and it seems natural. This would be a breaking change but a very necessary one (i think).

The marshal/unmarshal API should have an additional parameter `ctx: (Un-)MarshalContext` (or similar) like this:
```rust
MarshalContext <'a> {
    fds: &'a mut Vec<RawFd>,
    ... // maybe in the future other fields might be necessary here  
}
UnmarshalContext <'a> {
    fds: &'a [RawFd],
    ... // maybe in the future other fields might be necessary here  
}
```

This would allow for Marshal implementations for all `std::os::unix::io::IntoRawFd` types. It would also maybe allow for Unmarshal implementations for`std::os::unix::io::FromRawFd`. This would be unsafe though so this should be treated with care. Unmarshal for `u32` or `RawFd` is trivially possible though.

**Comment by KillingSpark** [comment]
This should wait until [this branch](https://github.com/KillingSpark/rustbus/tree/variant_trait_impl) for issue #35 is merged, because it means an extensive refactor of the core stuff which would maybe mean annoying merging.

**Comment by KillingSpark** [comment]
started doing the refactoring in [this branch](https://github.com/KillingSpark/rustbus/tree/improve_unixfd_handling)

**Comment by KillingSpark** [comment]
The strategy for now is:
1. Change all Marshal/Unmarshal trait impls to the new Context stuff (already done for Marshal). This does not change the behavior in any way. 
2. Change the Param implementation to the new Context stuff. This will also not change the behavior.
4. Up until now if any tests fail to run, I did something wrong while refactoring.
5. Change the way UnixFd in the traits and the Param implementations is (un-)marshalled
6. After this I will do some more testing, validate that fd transfere does indeed work (using the fd example and maybe extend that a bit)
7. It's merge time and time to bump the version yet again.

**Comment by KillingSpark** [comment]
This has been merged into master now 967bd62e2b1e6ab344413597058489cbc55ede3f
